### PR TITLE
feat: Track Fnames

### PR DESCRIPTION
**Pull request by adityapk00**
## Motivation

Track fname updates in the hub so users can be identified by their preferred fname.

## Change Summary

Update Hub to track `NameRegistry` events, persist them in RocksDB. A new message type `UserNameAdd` allows users to broadcast their preferred fname. 

## Merge Checklist

_Choose all relevant options below by adding an `x` now or at any time before submitting for review_

- [X] The title of this PR adheres to the [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/) standard
- [X] The PR has been tagged with the appropriate change type label(s) (i.e. documentation, feature, bugfix, or chore)
- [X] Changes to the protocol specification have been merged

## Additional Context
See https://github.com/farcasterxyz/hub/issues/128 for the feature spec. 

Since this is my first PR, I thought I'll send it for some early feedback. 

- This is missing a few test cases for some of the edge scenarios. 
- I'm having some issues signing the commits, so I might have to update the commits

### CODE REVIEW

**Comment by adityapk00** [comment]
Tagging @varunsrin @pfletcherhill

File: src/storage/sets/flatbuffers/signerStore.ts [hunk 1]
```diff
@@ -175,6 +181,24 @@ class SignerStore {
     this._eventHandler.emit('mergeContractEvent', event);
   }
 
+  /**
+   * Merges a NameRegistryEvent into the SignerStore, storing the causally latest event at the key:
+   * <RootPrefix:User><UserPostfix:NameRegistryEvent><fname>
+   */
+  async mergeNameRegistryEvent(event: NameRegistryEventModel): Promise<void> {
```
Comment (pfletcherhill) [inline src/storage/sets/flatbuffers/signerStore.ts:188]: This method should go in the UserDataStore. The `mergeIdRegistryEvent` method is in the SignerStore, because it relates to signers, but new NameRegistryEvents could cause UserNameAdd messages to be revoked, so it should be in the UserDataStore.

File: src/storage/sets/flatbuffers/userDataStore.ts [hunk 2]
```diff
@@ -133,11 +170,53 @@ class UserDataStore {
...
+  private async mergeNameAdd(message: UserNameAddModel): Promise<void> {
+    // TODO: Check that the user sending this message actually owns the fname
+
+    let tsx = await this.resolveUserNameMergeConflicts(this._db.transaction(), message);
+
+    // No-op if resolveMergeConflicts did not return a transaction
+    if (!tsx) return undefined;
+
+    // Revoke the name from the previous user
+    tsx = await this.revokeNameFromPreviousUser(tsx, message);
```
Comment (pfletcherhill) [inline src/storage/sets/flatbuffers/userDataStore.ts:200]: I don't think this should happen here. The UserDataStore should resolve conflicts between UserNameAdd messages (i.e. two from the same fid) but not affect data from other users. A UserNameAdd message should be revoked if/when:
* a NameRegistryEvent is merged that transfers ownership of the fname
* the fname expires

File: src/network/rpc/flatbuffers/client.ts [hunk 3]
```diff
@@ -216,6 +223,13 @@ class Client {
     );
   }
 
+  async getUserName(fid: Uint8Array): HubAsyncResult<UserNameAddModel> {
+    return this.makeUnaryUserNameRequest(
```
Comment (pfletcherhill) [inline src/network/rpc/flatbuffers/client.ts:227]: I think you can use the `makeUnaryMessageRequest` method here instead of making a new method

File: src/network/rpc/flatbuffers/client.ts [hunk 4]
```diff
@@ -312,6 +326,48 @@ class Client {
...
+            resolve(err(fromServiceError(e)));
+          } else if (response) {
+            resolve(ok(new NameRegistryEventModel(response)));
+          }
+        }
+      );
+    });
+  }
+
+  private makeUnaryUserNameRequest<RequestType, ResponseMessageType extends MessageModel>(
```
Comment (pfletcherhill) [inline src/network/rpc/flatbuffers/client.ts:350]: See comment above. You can get rid of this method in favor of `makeUnaryMessageRequest`.

**Review by pfletcherhill — CHANGES_REQUESTED** [review]
Okay, I went through most of the files. Let me know if you feel differently, but I think we can refactor `UserNameAdd` to be just another `UserDataAdd` with its own `UserDataType` value. We can re-use a lot of code that way.

**Review by pfletcherhill — APPROVED** [review]
Approving, though please address comments before merging
